### PR TITLE
Don't mutate children on DomNode 

### PR DESCRIPTION
**Pull request by eyebrowsoffire**
DomNode differs from the dart:html node in that dart:html does some magic under the hood to reflect changes made to the container to the underlying HTML node. DomNode does not do this, and its children are not mutable. Therefore use the normal JavaScript APIs to modify the tree.

Added a unit test that exercises this code path, which is when the chain of clip views changes length. This was the code path that was causing the crash described in the issue. The example in the issue reproduced this easily before these changes, and does not occur after.

Note that as part of updating the web engine to the new Dom API, @joshualitt is possibly planning on changing the type of the `children` field to be an immutable type (which would have caught this issue at compile time).

Fixes https://github.com/flutter/flutter/issues/104972

### CODE REVIEW

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 1]
```diff
@@ -260,9 +260,11 @@ class HtmlViewEmbedder {
     DomElement platformView,
     DomElement headClipView,
   ) {
-    int indexInFlutterView = -1;
-    if (headClipView.parent != null) {
-      indexInFlutterView = skiaSceneHost!.children.indexOf(headClipView);
```
Comment (joshualitt) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:265]: This is actually safe for now, but nice to get rid of if we decide we want to move off using `List` for `children`.

File: lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart [hunk 2]
```diff
@@ -732,7 +734,12 @@ class HtmlViewEmbedder {
 
   /// Deletes SVG clip paths, useful for tests.
   void debugCleanupSvgClipPaths() {
-    _svgPathDefs?.children.single.children.forEach(removeElement);
+    final DomElement? parent = _svgPathDefs?.children.single;
+    if (parent != null) {
+      for (DomNode? child = parent.lastChild; child != null; child = parent.lastChild) {
```
Comment (joshualitt) [inline lib/web_ui/lib/src/engine/canvaskit/embedded_views.dart:739]: We talked offline but:
```
while (parent.firstChild != null) {
  parent.removeChild(parent.firstChild)
}
```
is a bit more succinct. I'm okay with either.

**Review by joshualitt — CHANGES_REQUESTED** [review]
This looks good, modulo some minor feedback.

File: lib/web_ui/lib/src/engine/dom.dart [hunk 3]
```diff
@@ -149,6 +149,9 @@ extension DomNodeExtension on DomNode {
   DomElement? get parent => js_util.getProperty(this, 'parentElement');
   String? get text => js_util.getProperty(this, 'textContent');
   external DomNode? get parentNode;
+  external DomNode? get firstChild;
```
[the remark this answers is not in the document]
Comment (eyebrowsoffire) [reply lib/web_ui/lib/src/engine/dom.dart:152]: Yeah I hadn't merged main yet when I wrote this.

**Review by joshualitt — APPROVED** [review]

**Pull request merged**